src/objects.rs

```diff
@@ -211,8 +217,15 @@ impl ObjectData {
         // If the template attribute is there, we need to go fetch the template file
         let template = template
             .map(|template_path: String| {
-                let parent_dir = base_path.parent().ok_or(Error::PathIsNotFile)?;
-                let template_path = parent_dir.join(Path::new(&template_path));
+                // The base path may be a directory or a file
+                let template_path = {
+                    let path = PathBuf::from(template_path);
+                    if path.is_relative() {
+                        file_path_being_loaded.parent().unwrap().join(&path)
```

Comment:
Might there be a performance penalty from using `parent().unwrap()` in potentially many places? I personally liked that we were passing around the directory instead of the file path.

Reply:
Rather quite the opposite, we were previously prematurely obtaining the directory of the file being loaded before. Now, we only obtain that directory when it is actually required (when doing path manipulation for relative paths).

Reply:
Plus, ensuring that we are passing a file path as opposed to a file or directory path allows us to completely remove the "Path does not refer to a file" error in many places.

Reply:
> Rather quite the opposite, we were previously prematurely obtaining the directory of the file being loaded before. Now, we only obtain that directory when it is actually required (when doing path manipulation for relative paths).

Right, but previously it could happen exactly once per file, whereas now it happens once per file reference. Generally one can expect there to be a lot more file references than files, especially in template-heavy maps.

> Plus, ensuring that we are passing a file path as opposed to a file or directory path allows us to completely remove the "Path does not refer to a file" error in many places.

Yeah, I understand we need this to be consistent. I was just wondering whether it would be better to consistently pass the parent directory rather than consistently using the file path. Anyway, I also doubt it really matters.